w3lib/url.py

```diff
@@ -422,7 +422,9 @@ def parse_url(url, encoding=None):
     parsed url)
     """
     if isinstance(url, ParseResult):
-        return url
+        # we need to make sure we have unicode parts
+        # otherwise, things like IDNA encoding may fail on bytes
+        return ParseResult(*(to_unicode(part, encoding) for part in url))
```

Comment:
@kmike , regarding your first comment, scrapy is using [`to_native_string()` in the default lxml-based link extractor](https://github.com/scrapy/scrapy/blob/129421c7e31b89b9b0f9c5f7d8ae59e47df36091/scrapy/linkextractors/lxmlhtml.py#L59), which is not correct, and for which there's an open PR https://github.com/scrapy/scrapy/pull/1949